sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -470,3 +397,451 @@ object JoinReorderDPFilters extends PredicateHelper {
+
+    assert(pop.chromos.head.basicPlans.size == items.length)
+    pop.chromos.head.integratedPlan match {
+      case Some(joinPlan) => joinPlan.plan match {
+        case p @ Project(projectList, _: Join) if projectList != output =>
+          assert(topOutputSet == p.outputSet)
+          // Keep the same order of final output attributes.
+          Some(p.copy(projectList = output))
+        case finalPlan if !sameOutput(finalPlan, output) =>
+          Some(Project(output, finalPlan))
+        case finalPlan =>
+          Some(finalPlan)
+      }
+      case _ => None
+    }
+  }
+}
+
+/**
+ * A pair of parent individuals can breed a child with certain crossover process.
+ * With crossover, child can inherit gene from its parents, and these gene snippets
+ * finally compose a new [[Chromosome]].
+ */
+@DeveloperApi
+trait Crossover {
+
+  /**
+   * Generate a new [[Chromosome]] from the given parent [[Chromosome]]s,
+   * with this crossover algorithm.
+   */
+  def newChromo(father: Chromosome, mother: Chromosome) : Chromosome
+}
+
+/**
+ * This class implements the Genetic Edge Recombination algorithm. The algorithm generates
+ * a new traveling path by choosing the edges in certain order from the parents' edge table,
+ * where the edge table contains links of vertexes of the parents' traveling paths.
+ *
+ * Here's an example. Suppose we have two traveling paths on a graph,
+ *      I, [A B C D E F]
+ *     II, [B D C A E F]
+ *
+ * The algorithm works as follows,
+ *  1. find the links of each vertex, then we have an 'Edge Table'
+ *      A: B F C E
+ *      B: A C D F
+ *      C: B D A
+ *      D: C E B
+ *      E: D F A
+ *      F: A E B
+ *  2. from one vertex, say A, choose one of it's neighbours as it's new neighbour, say F, we have
+ *      {A F}
+ *  3. then choose one from F's neighbours(note that vertex that has been chosen should not
+ *     be chosen again),
+ *      {A F E}
+ *  4. go on,
+ *      {A F E D}
+ *  5. go on,
+ *      {A F E D B}
+ *  6. go on,
+ *      {A F E D B C}
+ *  Note if the procedure ends before all of vertexes been chosen, start from another vertex that
+ *  has not been chosen, and go on the procedure.
+ *
+ * For more information about the Genetic Edge Recombination,
+ * see "Scheduling Problems and Traveling Salesmen: The Genetic Edge
+ * Recombination Operator" by Darrell Whitley et al.
+ * https://dl.acm.org/citation.cfm?id=657238
+ */
+object EdgeRecombination extends Crossover {
+
+  def genEdgeTable(father: Chromosome, mother: Chromosome) : Map[JoinPlan, Seq[JoinPlan]] = {
+    val fatherTable = father.basicPlans.map(g => g -> findNeighbours(father.basicPlans, g)).toMap
+    val motherTable = mother.basicPlans.map(g => g -> findNeighbours(mother.basicPlans, g)).toMap
+
+    fatherTable.map(entry => entry._1 -> (entry._2 ++ motherTable(entry._1)))
+  }
+
+  def findNeighbours(genes: Seq[JoinPlan], g: JoinPlan) : Seq[JoinPlan] = {
```

Comment:
An elegant approach is to calculate `(i-1+seq.length) % seq.length`/`(i+1+seq.length) % seq.length`. No need for a single `if`.

Reply:
This can absolutely be a one-liner, consider `flatMap`?

Reply:
Sorry @yeshengm , I didn't get you on "one-liner, consider `flatMap`". However I updated the code, and it seems not as cumbersome as before.

Reply:
Consider this `x.zipWithIndex.flatMap { case (v, idx) => Seq(v -> x((idx+1)%v.length), v->x((idx-1+v.length())%v.length)) }`. Definitely too long for a line, but you get the idea.

Reply:
Updated.